dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -91,12 +95,20 @@
     public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
         // Optimistically avoid the queue in the fast path.
         // Queuing adds contention between threads and should be avoided unless we need to shed load.
-        Optional<ListenableFuture<Response>> maybeResult = delegate.maybeExecute(endpoint, request);
-        if (maybeResult.isPresent()) {
-            ListenableFuture<Response> result = maybeResult.get();
-            numRunningRequests.incrementAndGet();
-            result.addListener(this::onCompletion, DIRECT);
-            return result;
+        if (queueSizeEstimate.get() <= 0) {
+            Optional<ListenableFuture<Response>> maybeResult = delegate.maybeExecute(endpoint, request);
+            if (maybeResult.isPresent()) {
+                ListenableFuture<Response> result = maybeResult.get();
+                numRunningRequests.incrementAndGet();
+                result.addListener(this::onCompletion, DIRECT);
+                return result;
+            }
+        }
```

Comment:
this is a behaviour change right? where previously requests could jump the queue if the channel happened to open up, but they're now more FIFO?

Reply:
That's correct, it's why the tests changed. I avoided checking the queue size using the old implementation because LinkedBlockingQueue acquires locks to check the size, but this is a quick volatile read.